vspec/model/constants.py

```diff
@@ -10,127 +10,159 @@
+import pkg_resources
+from typing import (
+    Sequence, Type, TypeVar, Optional, Dict, Tuple, NamedTuple, Iterator,
+)
+
+import yaml
+
+from vspec.model import CONFIG_PATH, NON_ALPHANUMERIC_WORD
 
 
 T = TypeVar("T")
 
 
-class EnumMetaWithReverseLookup(EnumMeta):
-    """This class extends EnumMeta and adds:
+class VSSConstant(NamedTuple):
+    label: str
+    value: str
+    description: Optional[str] = None
+    domain: Optional[str] = None
+
+
+def dict_to_constant_config(name: str, info: Dict[str, str]) -> Tuple[str, VSSConstant]:
+    label = info['label']
+    label = NON_ALPHANUMERIC_WORD.sub('', label).upper()
+    description = info.get('description', None)
+    domain = info.get('domain', None)
+    return label, VSSConstant(info['label'], name, description, domain)
+
+
+def iterate_config_members(config: Dict[str, Dict[str, str]]) -> Iterator[Tuple[str, VSSConstant]]:
+    for u, v in config.items():
+        yield dict_to_constant_config(u, v)
+
+
+class VSSEnumMeta(type):
+    """This class defines the enumeration behavior for vss:
+     - Access through Class.ATTRIBUTE
      - from_str(str): reverse lookup
      - values(): sequence of values
     """
-    def __new__(typ, *args, **kwargs):
-        cls = super().__new__(typ, *args, **kwargs)
-        if not hasattr(cls, "__reverse_lookup__"):
+
+    def __new__(mcs, cls, bases, classdict):
+        cls = super().__new__(mcs, cls, bases, classdict)
+
+        if not hasattr(cls, '__reverse_lookup__'):
             cls.__reverse_lookup__ = {
-                v.value: v for v in cls.__members__.values()
+                v.value: v for k, v in cls.__config__.items()
             }
-        if not hasattr(cls, "__values__"):
-            cls.__values__ = tuple(v.value for v in cls.__members__.values())
+        if not hasattr(cls, '__values__'):
+            cls.__values__ = set(cls.__reverse_lookup__.keys())
+
         return cls
 
+    def __getattr__(cls, key: str) -> str:
+        try:
+            return cls.__config__[key]
+        except KeyError as e:
+            raise AttributeError(
+                f"type object '{cls.__name__}' has no attribute '{key}'"
+            ) from e
+
+    def from_str(cls: Type[T], value: str) -> T:
+        return cls.__reverse_lookup__[value]
+
+    def values(cls: Type[T]) -> Sequence[str]:
+        return cls.__values__
+
+
+class VSSEnumMetaConfigurable(VSSEnumMeta):
+    def __new__(mcs, cls, bases, classdict):
+        cls = super().__new__(mcs, cls, bases, classdict)
+
+        with open(CONFIG_PATH, mode='r') as config_file:
+            yaml_config = yaml.safe_load(config_file)
+        config = yaml_config.get(classdict['__config_name__'], {})
+        cls.add_config(config)
+        return cls
+
+    def add_config(cls, config: Dict[str, Dict[str, str]]):
+        for k, v in iterate_config_members(config):
+            cls.__config__[k] = v
+            cls.__reverse_lookup__[v.value] = v
+            cls.__values__.add(v.value)
```

Comment:
if `__values__` is converted to a list, you can cheat by using `cls.__reverse_lookup` check first:

```py
if v.value not in cls.__reverse_lookup__:
    cls.__reverse_lookup__[v.value] = v
    cls.__values__.append(v.value)
```

IF you can allow replacements, then you move the `cls.__reverse_lookup__[v.value] = v` *after* the `if` ends.

But this while `add_config` should be in the base metaclass

Reply:
Done